### ISSUE TITLE
Is the ArcFace loss correctly implemented?

### ISSUE THREAD

**Issue by jolastar** [TMaysGGS/MobileFaceNet-Keras#4]
I haven't tested it yet, but looking at the code - there is a line which looks very fishy to me:

https://github.com/TMaysGGS/MobileFaceNet-Keras/blob/7ee0dc91bd370a7a8d81ccafcaa5c7de8cb2106d/Tools/Keras_custom_layers.py#L142

I know the implementation comes from [ewrfcas/Machine-Learning-Toolbox](https://github.com/ewrfcas/Machine-Learning-Toolbox/blob/master/loss_function/ArcFace_loss.py) and the line is identical there.

I suspect that the line should not be there at all. Or maybe there was some kind of numeric issue that made the author want to do `y_mask += K.epsilon()` or `y_mask -= K.epsilon()`?

In any case, as it stands, the code causes the softmax to be applied uniformly to all logits, whether they match the label or not (because `y_mask` and `inv_mask` are constants) and the layer boils down to a regular softmax layer with scaling.

What do you think?

**Comment by TMaysGGS** [comment]
Hi @jolastar ,

Thanks for the check. Indeed "=+" will make y_mask a constant and I also think this line should not be there since K.epsilon is default to 1e-7, which is usually used to avoid a ZERO denominator while y_mask seems not to be one. And I checked some other TensorFlow implementation. It seems they just do not have this line. 

Later I will check the official MXNet implementation of Arcface Loss and revise this. Thanks again for the check.

**Comment by jolastar** [comment]
In the "official MXNet implementation", note that there are conflicting implementations of the ArcFace loss as I pointed out in [this issue](https://github.com/deepinsight/insightface/issues/1011).

And I'm curious: which TensorFlow implementation do you recommend as reference?

**Comment by TMaysGGS** [comment]
> In the "official MXNet implementation", note that there are conflicting implementations of the ArcFace loss as I pointed out in [this issue](https://github.com/deepinsight/insightface/issues/1011).
> 
> And I'm curious: which TensorFlow implementation do you recommend as reference?

You can search "Third-party Re-implementation" in the official page of InsightFace [https://github.com/deepinsight/insightface] and there are four TensorFlow re-implementations among the eight. I am a little busy recently so I just revised the ArcFace Loss Layer according to the tf version. I can check the official MXNet and your issue there later some time. 

BTW, have you succeeded getting a relatively good model? My hardware is very limited so I have not got one yet.

**Comment by TMaysGGS** [comment]
> In the "official MXNet implementation", note that there are conflicting implementations of the ArcFace loss as I pointed out in [this issue](https://github.com/deepinsight/insightface/issues/1011).
> 
> And I'm curious: which TensorFlow implementation do you recommend as reference?

Hi, I just modified the ArcFace Loss Layer by just deleting the line "y_mask =+ K.epsilon()" and retrained my model for a couple of iterations (not epochs). It seems the loss never decreases so it may be not right for justing transfoming the codes in TensorFlow to Keras. I will try to debug continuously.

**Comment by TMaysGGS** [comment]
> > In the "official MXNet implementation", note that there are conflicting implementations of the ArcFace loss as I pointed out in [this issue](https://github.com/deepinsight/insightface/issues/1011).
> > And I'm curious: which TensorFlow implementation do you recommend as reference?
> 
> Hi, I just modified the ArcFace Loss Layer by just deleting the line "y_mask =+ K.epsilon()" and retrained my model for a couple of iterations (not epochs). It seems the loss never decreases so it may be not right for justing transfoming the codes in TensorFlow to Keras. I will try to debug continuously.

After I tried again for training the model using both the modified arcface loss & the softmax loss for a smaller data set, both the losses are decreasing significantly and the outcomes seems right. The former attempt that showed the loss not decreasing had not happened again. So I close this issue now.
